ts-todo-list

todnks opened the pull request.

todnks committed:
feat: 초기설정

todnks committed:
chore: resetcss삭제

todnks committed:
feat: localstorage에 데이터추가기능

todnks committed:
feat: 목록선택추가

todnks committed:
refactor 로직변경

todnks committed:
fix: 오타수정

todnks committed:
feat: id추가list출력로직변경

todnks committed:
refactor: 정의되지않은타입과type오류수정
Appkeycod key로 변경, localStorage key값 따로 지정
state값타입지정 countContainer이름변경

minhyeok0328 commented on src/App.ts:
```diff
@@ -0,0 +1,61 @@
+    const input: HTMLInputElement = document.querySelector('.new-todo')!;
+    input?.addEventListener('keypress', (event: KeyboardEvent) => {
+
+      let storageData: todoListData[] = JSON.parse(localStorage.getItem(keys.key)!);
+
+      if (event.key !== "Enter" || !input.value) return;
+
+      const insertData: todoListData = {
+        id: new Date().getTime(),
+        title: input.value,
+        completed: false
+      }
+
+      if (storageData !== null) storageData.push(insertData);
+      if (storageData === null) storageData = [insertData];
+
+      localStorage.setItem(keys.key, JSON.stringify(storageData));
```
여기서 localStorage를 계속 반복적으로 호출하고 있는데 저걸 추상화 시켜서 따로 분리해 보는 건 어떨까?

예를들어

```
class Storage {
    private readonly storage: Storage;
    private readonly key: string;
    
    constructor(key: string, storage: Storage = localStorage) {
        this.key = key;
        this.storage = storage;
    }

    getItem() {}
    setItem() {}
    removeItem() {}
}


const storage = new Storage('todo');

storage.setItem([]);
```

이렇게 해 두고 import 해서 사용하면 계속 저렇게 constants에서 key 상수를 가져와서 계속 넣어줄 필요가 없을 것 같네

minhyeok0328 commented on src/constants/index.ts:
```diff
@@ -0,0 +1,3 @@
+export const keys = {
+  key: 'todos'
+};
```
변수명이 그냥 keys 라서 좀 더 명확하게 어떤 변수인지 이름을 바꿔주면 좋겠다~

ex) `export const TODO_STORAGE_KEY = 'todos';`

minhyeok0328 commented on src/container/listContainer.ts:
```diff
@@ -0,0 +1,35 @@
+export class listContainer extends component {
+
+  setup() {
+    this.state = JSON.parse(localStorage.getItem(keys.key)!);
```
`this.state = JSON.parse(localStorage.getItem(keys.key)!);`

이렇게 하면 JSON.parse가 실패했을 때 아래 기능들도 영향을 받을 수 있으니 저렇게 ! 를 붙이면 좀 안좋을 수 있다!

minhyeok0328 commented on src/container/listContainer.ts:
```diff
@@ -0,0 +1,35 @@
+  setEvent() {
+
+    const destroyBtn = document.querySelectorAll(".destroy");
+    destroyBtn.forEach(element => {
+      element.addEventListener("click", () => {
+        const data = this.state?.filter(data => String(data.id).indexOf(element.id));
+        localStorage.setItem(keys.key, JSON.stringify(data));
+        new listContainer(document.querySelector('#todo-list')!);
+      })
+    });
```
삭제 버튼 전체를 forEach 돌면서 이벤트를 걸어주고 있는데 이러면 저 버튼이 많아짐에 따라 이벤트도 점점 많이 걸리기 때문에

`**이벤트 위임**` 이라는 것도 한 번 찾아보면 좋을 것 같다~

minhyeok0328 commented on src/core/component.ts:
```diff
@@ -0,0 +1,24 @@
+interface todoListData {
+  id: number,
+  title: string,
+  completed: boolean
+}
```
todoListData 라는 interface가 계속 반복적으로 나오고 있는데 공통으로 사용하는 interface라면 따로 분리해서 각 파일에서 import 해서 사용하면 더 좋을 것 같다~

minhyeok0328 commented on src/core/component.ts:
```diff
@@ -0,0 +1,24 @@
+}
+export class component {
+  $element: HTMLElement;
+  state?: todoListData[];
```
component 클래스가 component를 정의할 때 공통으로 사용하는 클래스 같은데

`state?: todoListData[];`

이렇게 state에 todoListData 라는 interface를 넣어두면 여기서 다른 기능을 추가하기 어렵지 않을까~
차라리 제네릭으로 받는 방법도 괜찮아 보인다~

minhyeok0328 commented on src/App.ts:
```diff
@@ -0,0 +1,61 @@
+      if (storageData === null) storageData = [insertData];
+
+      localStorage.setItem(keys.key, JSON.stringify(storageData));
+      this.onMounted();
```
core/component.js 파일 보니까 setEvent 끝나고 onMounted도 실행시켜 주던데 굳이 이렇게 해준 이유가 따로 있니? 이건 그냥 궁금해서 물어보는거야

minhyeok0328 reviewed (commented):
아직 typescript 사용이 익숙하지는 않은 것 같으니 공부를 계속 해보면 좋을 것 같다~

todnks committed:
chore: eslint파일수정, rules추가

todnks committed:
refactor: class 형식을 functional형식으로 변경
자주쓰는 타입은 따로 정의

todnks committed:
feat: 컴포넌트 생성하며 관리더 쉽게 생성

todnks committed:
feat: 자주쓰는 기능들 따로 core에 생성
selector와storage기능 따로 빼서 상태관리 편의하게 추가

todnks committed:
feat: list체크 기능추가
기능추가하면서 input checked설정추가 및 li class 이름 설정

todnks committed:
feat: List수정추가
dbclick으로 editing클래스부여후
keypress로 수정사항 수정결정및 setitem으로 재랜더

todnks committed:
refactor: 키값이름 변경

todnks committed:
feat: a태그 컴포넌트화
component화시켜서 관리및 유지보수 쉽게 만듬

todnks committed:
feat: router추가
해야할일과완료한일이 있어 router를 추가했습니다. routing에서 랜더를 실행시킴 push로 존재하는주소인지 확인후push함

todnks committed:
feat: 리스트를 url에 맞게 상태변환
router를 이용해서 url체크후에 맞는 현재url에 맞게 리스트출력

todnks committed:
feat: 수정중 esc입력시 취소

todnks committed:
feat: 브라우저가 처음열릴때 hash가 없을경우 기본 리스트 출력

todnks committed:
refactor: 에러 페이지 기본페이지로변경

todnks committed:
Update README.md

minhyeok0328 commented on src/App.ts:
```diff
@@ -0,0 +1,63 @@
+    newTodoList.push(insertData);
+    setItem(STORAGE_KEY.TODO_KEY, newTodoList);
+    input.value = '';
+  });
```
이 부분은 그냥 keypress 이벤트를 등록하는 부분이라 App.ts 에서 몰라도 되는 내용 아닐까 싶네
저 함수를 따로 분리해서 관리하고 사용할 때 import 해서 가져오면 좀 더 좋을 것 같아

minhyeok0328 commented on src/App.ts:
```diff
@@ -0,0 +1,63 @@
+import { addEvent } from './core/Render';
+export default function App() {
+
+  let storageData: todoListData[] = getItem(STORAGE_KEY.TODO_KEY);
```
`storageData` 변수 보니까 딱히 변경될 일이 없던데 이럴때는 const 로 선언하는게 좋아

minhyeok0328 commented on src/App.ts:
```diff
@@ -0,0 +1,63 @@
+    input.value = '';
+  });
+
+  const hashName = location.hash.replace('#', '');
+  const routes = {
+    '': storageData,
+    '/': storageData,
+    '/active': storageData.filter((data: todoListData) => data.completed === false),
+    '/completed': storageData.filter((data: todoListData) => data.completed === true),
+  };
+  const findroutes = Object.keys(routes).findIndex((key) => key === hashName);
+  const ListData = Object.values(routes)[findroutes];
```
이 부분은 라우팅 관련 된 로직 같은데 이런거는 App.ts 말고 따로 라우팅 관련 파일을 만들어서 거기서 관리하는게 좋아 보니까 router 폴더가 따로 있던데 거기서 관련된 로직을 모아서 관리하면 다른 사람이 코드 볼 때도 router 관련된 코드는 router 폴더 안에 다 있으니까 보기 편하겠지?

minhyeok0328 commented on src/components/Title.ts:
```diff
@@ -0,0 +1,3 @@
+export default function Title(props: string) {
+  return `<h1>${props}</h1>`;
+}
```
`props` 라는 변수 명 보다는 좀 더 명확하게 변경해보는건 어떨까?

가령 `text` 라던지 `pageTitle` 이라던지 보면 바로 알 수 있게끔

minhyeok0328 commented on src/components/TodoItem.ts:
```diff
@@ -0,0 +1,67 @@
+  addEvent('.todo-list', 'click', ({ target }: HTMLElement) => {
+
+    if (target.className === 'destroy') {
+      const data: todoListData['id'] = getItem(STORAGE_KEY.TODO_KEY).filter((data: todoListData) => data.id.indexOf(target.dataset.id));
```
보니까 전체적으로 getItem 함수에 `STORAGE_KEY.TODO_KEY` 라는 키 값을 넘겨주는데 이러면 사용할 때 마다 계속 넘겨줘여 하니까 저 함수 쪽에서 키 값을 관리할 수 있도록 해보는건 어떨까? getItem 함수를 사용하는 경우에는 굳이 키 값을 몰라도 되게끔 하는거지

minhyeok0328 commented on src/components/TodoItem.ts:
이 파일 전체적으로 너무 많은 내용이 담겨있어
addEvent 해주는 부분은 두고 안에 있는 함수들을 바깥으로 따로 분리해서 관리하면 좋을 것 같아

minhyeok0328 commented on src/core/Router.ts:
todolist는 페이지도 하나밖에 없어서 페이지 이동이 따로 있는 것도 아닌데 굳이 라우터를 만들어서 한 이유가 따로 있을까?

minhyeok0328 commented on src/core/BaseStorage.ts:
BaseStorage 라는 함수를 잘 이용하면 getitem, setItem을 사용할 때 굳이 key 값을 안 넘겨도 되지 않을까?

minhyeok0328 commented on src/core/BaseStorage.ts:
```diff
@@ -0,0 +1,25 @@
+}
+
+function BaseStorage() {
+  const getItem = (key: stroage['key']) => {
```
그냥 `key: string` 이렇게 해줘도 될 것 같아

minhyeok0328 reviewed (commented):
일단.. 로직들이 각 파일, 컴포넌트 들에 작성되어 있어서 전체적으로 가독성이 그닥 좋지는 않아
따로 함수나 관련 로직들을 모아서 관리하는게 좋아 보이고 변수명 들도 전체적으로 개선이 필요한 것 같아

단순히 props 로 하는 것 보다는 봤을 때 이건 어떤 값이 담기는 변수구나! 라고 알 수 있게 끔 네이밍을 변경해주면 좋겠어

구글번역기 돌려보면서 네이밍 연습도 해보면 좋겠네

그리고 todolist 만드는데 라우터가 굳이 필요할까..? 라는 생각이 들어
그래서 전체적으로 코드를 정리정돈 하는 게 우선 필요할 것 같아

코드 중간중간에 개행을 적절하게 해 주는 것도 필요하고

todnks committed:
refactor: Storage로직변경
key를 파일마다 import하지 않고 baseStorage에서 관리. baseStorage에 todoItem관리 로직추가

todnks committed:
Merge branch 'feature/test' of https://github.com/todnks/ts-todo-list into feature/test

todnks committed:
style: props이름을 좀더 가독성있게 변경

todnks committed:
refactor: addevent안에 있는 함수 따로 관리

todnks committed:
refactor: input 오류

todnks committed:
refactor: hashName에 따른 list출력 로직 따로 관리

todnks committed:
refactor: router삭제하고 리스트조건에맞게 출력하는 이벤트추가

todnks committed:
refactor: BaseStorage수정
기존에있던 getTodoItem,setTodoItem을 없애고 BaseStorage를 실행할때 key값과 storage값을 입력하는 방식으로 변경

todnks committed:
refactor: newtodoData로직변경

todnks committed:
refactor: Input에서 newTodoData제거

todnks committed:
refactor: 글수정 로직변경
useTodoItem에서 editingBox TodoItem으로 옮김 editingTodoData를 dataId와 data를 받아서 실행으로 변경

minhyeok0328 committed:
Merge branch 'main' into feature/test
